kernel/aster-nix/src/net/socket/vsock/stream/socket.rs

```diff
@@ -0,0 +1,379 @@
+            }
+        };
+        let remote_addr = VsockSocketAddr::try_from(sockaddr)?;
+        let local_addr = init.bound_addr();
+
+        if let Some(addr) = local_addr {
+            if addr == remote_addr {
+                return_errno_with_message!(Errno::EINVAL, "try to connect to self is invalid");
+            }
+        } else {
+            init.bind(VsockSocketAddr::any_addr())?;
+        }
+
+        let connecting = Arc::new(Connecting::new(remote_addr, init.bound_addr().unwrap()));
+        let vsockspace = VSOCK_GLOBAL.get().unwrap();
+        vsockspace
+            .connecting_sockets
+            .lock_irq_disabled()
+            .insert(connecting.local_addr(), connecting.clone());
+
+        // Send request
+        vsockspace
+            .driver
+            .lock_irq_disabled()
+            .request(&connecting.info())
+            .map_err(|e| Error::with_message(Errno::EAGAIN, "can not send connect packet"))?;
+
+        // wait for response from driver
+        // TODO: add timeout
+        let poller = Poller::new();
+        if !connecting
+            .poll(IoEvents::IN, Some(&poller))
+            .contains(IoEvents::IN)
+        {
+            poller.wait()?;
+        }
+        vsockspace
+            .connecting_sockets
+            .lock_irq_disabled()
+            .remove(&connecting.local_addr())
+            .unwrap();
+
+        let connected = Arc::new(Connected::from_connecting(connecting));
+        *self.status.write() = Status::Connected(connected.clone());
+        // move connecting socket map to connected sockmap
+        vsockspace
+            .connected_sockets
+            .lock_irq_disabled()
+            .insert(connected.id(), connected);
+
+        Ok(())
+    }
+
+    fn listen(&self, backlog: usize) -> Result<()> {
+        let init = match &*self.status.read() {
+            Status::Init(init) => init.clone(),
+            Status::Listen(_) => {
+                return_errno_with_message!(Errno::EINVAL, "The socket is already listened");
+            }
+            Status::Connected(_) => {
+                return_errno_with_message!(Errno::EISCONN, "The socket is already connected");
+            }
+        };
+        let addr = init.bound_addr().ok_or(Error::with_message(
+            Errno::EINVAL,
+            "The socket is not bound",
+        ))?;
+        let listen = Arc::new(Listen::new(addr, backlog));
+        *self.status.write() = Status::Listen(listen.clone());
+
+        // push listen socket into vsockspace
+        VSOCK_GLOBAL
+            .get()
+            .unwrap()
+            .listen_sockets
+            .lock_irq_disabled()
+            .insert(listen.addr(), listen);
+
+        Ok(())
+    }
+
+    fn accept(&self) -> Result<(Arc<dyn FileLike>, SocketAddr)> {
+        if self.is_nonblocking() {
+            self.try_accept()
+        } else {
+            wait_events(self, IoEvents::IN, || self.try_accept())
```

Comment:
Implement `wait_events` on `VsockStreamSocket`, and use `self.wait_events`, instead.